### PR TITLE
deprecate flush_logs_every_n_steps on Trainer

### LINKED ISSUE #8991: Deprecate `flush_logs_every_n_steps` on Trainer
## 🚀 Feature

Deprecate `flush_logs_every_n_steps` from Trainer and make it available as a parameter to loggers that have this capability.

### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:

- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

Flushing should be considered an internal implementation detail of each logger. For example, TensorBoard automatically flush logs after a given amount of time (`flush_secs`).

Currently, flushing logs is configured through Trainer, which seems like the wrong level of abstraction. Setting `flush_logs_every_n_steps` given a TensorBoard logger doesn’t actually flush to disk, but calls `log_metrics`, which can be misleading to the user.

Prior issue: https://github.com/PyTorchLightning/pytorch-lightning/issues/4664

### Pitch

Deprecate `flush_logs_every_n_steps` from Trainer, and move it to the init for logger classes that support this functionality (e.g. CSVLogger). 
The logger connector already passes in the step information (`self.trainer.logger.agg_and_log_metrics(scalar_metrics, step=step)`) so we can move the flushing logic to a utility function.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by edward-io**
## What does this PR do?

- Deprecates flush_logs_every_n_steps
- Moves flush_logs_every_n_steps init to individual loggers (only CSV logger has the ability to flush afaict)

Fixes #8991

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/csv_logs.py [hunk 1]
```diff
@@ -197,6 +200,8 @@ def log_hyperparams(self, params: Union[Dict[str, Any], Namespace]) -> None:
     def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) -> None:
         metrics = self._add_prefix(metrics)
         self.experiment.log_metrics(metrics, step)
+        if step is not None and (step + 1) % self._flush_logs_every_n_steps == 0:
+            self.save()
```
Comment (ananthsub) [inline pytorch_lightning/loggers/csv_logs.py:204]: if `step is None` then we'll never trigger saving? 
should we track how many times `log_metrics` is called with a separate counter, and then use that for determining when to save?

Comment (edward-io) [reply pytorch_lightning/loggers/csv_logs.py:204]: Right, saving would only be done at the end when `finalize` is being called. I'm not sure about the semantics of having an internal counter, since it would be different from the `global_step` of the trainer. What do you think?

**Comment by edward-io** [comment]
I'm not sure why the doc tests are failing. There aren't any obvious error messages, only some warnings. Building them locally on a laptop works for me.

**Comment by awaelchli** [comment]
I think that's because you commited submodule changes in _notebooks. Make sure to remove that again.

**Comment by awaelchli** [comment]
TensorBoardLogger also has "flushing" implemented as part of save(). Save is what is called in the training loop when `should_flush_logs=True`. And on the summary writer for tensorboard there is also a flush method but we only call it in `finalize()`. So confusing. I guess it's good that this gets deprecated.

File: pytorch_lightning/loggers/tensorboard.py [hunk 2]
```diff
@@ -100,6 +103,7 @@ def __init__(
         self._default_hp_metric = default_hp_metric
         self._prefix = prefix
         self._fs = get_filesystem(save_dir)
+        self._flush_logs_every_n_steps = flush_logs_every_n_steps
```
Comment (ananthsub) [inline pytorch_lightning/loggers/tensorboard.py:106]: @awaelchli @edward-io isn't this redundant with `max_queue` and `flush_secs` in the tensorboard summary writer constructor? https://pytorch.org/docs/stable/_modules/torch/utils/tensorboard/writer.html#SummaryWriter

Comment (edward-io) [reply pytorch_lightning/loggers/tensorboard.py:106]: It is redundant (now we have three different ways to flush logs!). It depends how strongly if we want to maintain backwards compatibility by keeping the old behavior of flushing per step.

Comment (ananthsub) [reply pytorch_lightning/loggers/tensorboard.py:106]: IMO it's better to describe how someone can configure the existing tensorboard logger for this. Given kwargs are passed through from the constructor to the experiment writer, we could demonstrate how max queue and flush_secs could be used to configure this behavior. I prefer relying on the libraries existing APIs rather than establishing something on top that's lightning specific especially because all these loggers have different APIs .

@awaelchli what do you prefer?

Comment (awaelchli) [reply pytorch_lightning/loggers/tensorboard.py:106]: ok I guess for the constructor that's fair.

File: pytorch_lightning/loggers/tensorboard.py [hunk 3]
```diff
@@ -73,6 +73,8 @@ class TensorBoardLogger(LightningLoggerBase):
         default_hp_metric: Enables a placeholder metric with key `hp_metric` when `log_hyperparams` is
             called without a metric (otherwise calls to log_hyperparams without a metric are ignored).
         prefix: A string to put at the beginning of metric keys.
+        sub_dir: Optional subdirectory to store logs.
```
Comment (awaelchli) [inline pytorch_lightning/loggers/tensorboard.py:76]: this is duplicated

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -214,6 +214,10 @@ def __init__(
 
             flush_logs_every_n_steps: How often to flush logs to disk (defaults to every 100 steps).
 
+                .. deprecated:: v1.5
+                    ``flush_logs_every_n_steps`` has been deprecated in v1.5 and will be removed in v1.7.
+                    Please pass ``flush_logs_every_n_steps`` directly to the Logger instead.
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:219]:
```suggestion
                    Please configure flushing directly in the logger instead.
```
as discussed in thread below, there is no intention to standardize this in the logger api

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 5]
```diff
@@ -44,11 +45,17 @@ def __init__(self, trainer: "pl.Trainer", log_gpu_memory: Optional[str] = None)
...
-        flush_logs_every_n_steps: int,
+        flush_logs_every_n_steps: Optional[int],
         log_every_n_steps: int,
         move_metrics_to_cpu: bool,
     ) -> None:
         self.configure_logger(logger)
+        if flush_logs_every_n_steps is not None:
+            rank_zero_deprecation(
+                f"Setting `Trainer(flush_logs_every_n_steps={flush_logs_every_n_steps})` is deprecated in v1.5 "
+                "and will be removed in v1.7. Please pass `flush_logs_every_n_steps` to the logger instead."
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:56]:
```suggestion
                "and will be removed in v1.7. Please configure flushing directly in the logger instead."
```
see comment below

**Review by awaelchli — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 6]
```diff
@@ -44,11 +45,17 @@ def __init__(self, trainer: "pl.Trainer", log_gpu_memory: Optional[str] = None)
...
         log_every_n_steps: int,
         move_metrics_to_cpu: bool,
     ) -> None:
         self.configure_logger(logger)
+        if flush_logs_every_n_steps is not None:
+            rank_zero_deprecation(
+                f"Setting `Trainer(flush_logs_every_n_steps={flush_logs_every_n_steps})` is deprecated in v1.5 "
+                "and will be removed in v1.7. Please configure flushing in the logger instead."
+            )
+        flush_logs_every_n_steps = 100  # original default parameter
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:58]: since it's deprecated and not removed entirely, should the logger still consider this value if not provided in the logger explicitly?

Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:58]: @edward-io is an indentation issue? idd you mean to apply this if `flush_logs_every_n_steps is None` ?

**Review by rohitgr7 — APPROVED** [review]
LGTM!

**Review by tchaton — APPROVED** [review]
LGTM !
